File: golang/receipt.go
package novax

import (
	"net/http"
	"net/url"
	"strconv"

	"github.com/shopspring/decimal"
)

// ReceiptProtocolsRequest lists supported chains.
// ReceiptType: 1 = dynamic address, 6 = fixed address (defaults to 1).
type ReceiptProtocolsRequest struct{ ReceiptType int }

func (r *ReceiptProtocolsRequest) Method() string     { return http.MethodGet }
func (r *ReceiptProtocolsRequest) Path() string       { return "/pay/v3/protocols" }
func (r *ReceiptProtocolsRequest) Extra() http.Header { return nil }
func (r *ReceiptProtocolsRequest) Body() any          { return nil }
func (r *ReceiptProtocolsRequest) Query() url.Values {
	t := r.ReceiptType
	if t == 0 {
		t = 1
	}
	return url.Values{"type": {strconv.Itoa(t)}}
}

// ReceiptAddressRequest looks up a fixed receipt address.
type ReceiptAddressRequest struct {
	Protocol             string
	SmartContractAddress string
	CompanyUserID        string
}

func (r *ReceiptAddressRequest) Method() string     { return http.MethodGet }
func (r *ReceiptAddressRequest) Path() string       { return "/pay/v3/receipt/address" }
func (r *ReceiptAddressRequest) Extra() http.Header { return nil }
func (r *ReceiptAddressRequest) Body() any          { return nil }
func (r *ReceiptAddressRequest) Query() url.Values {
	q := url.Values{}
	setQ(q, "protocol", r.Protocol)
	setQ(q, "smartContractAddress", r.SmartContractAddress)
	setQ(q, "companyUserId", r.CompanyUserID)
	return q
}

// ReceiptOrdersRequest queries receipt orders.
type ReceiptOrdersRequest struct {
	ReceiptOrderIDs string
	OrderStatus     *int
	Status          *int
}

func (r *ReceiptOrdersRequest) Method() string     { return http.MethodGet }
func (r *ReceiptOrdersRequest) Path() string       { return "/pay/v3/receipt/orders" }
func (r *ReceiptOrdersRequest) Extra() http.Header { return nil }
func (r *ReceiptOrdersRequest) Body() any          { return nil }
func (r *ReceiptOrdersRequest) Query() url.Values {
	q := url.Values{}
	setQ(q, "receiptOrderIds", r.ReceiptOrderIDs)
	if r.OrderStatus != nil {
		q.Set("orderStatus", strconv.Itoa(*r.OrderStatus))
	}
	if r.Status != nil {
		q.Set("status", strconv.Itoa(*r.Status))
	}
	return q
}

// ReceiptAddOrderRequest creates a receipt order.
type ReceiptAddOrderRequest struct {
	Protocol             string
	Currency             string
	SmartContractAddress string
	CompanyUserID        string
	ReceiptOrderID       string
	CurrencyNumber       decimal.Decimal
	CallBackURL          string
}

func (r *ReceiptAddOrderRequest) Method() string     { return http.MethodPost }
func (r *ReceiptAddOrderRequest) Path() string       { return "/pay/v3/receipt/order/add" }
func (r *ReceiptAddOrderRequest) Extra() http.Header { return nil }
func (r *ReceiptAddOrderRequest) Query() url.Values  { return nil }
func (r *ReceiptAddOrderRequest) Body() any {
	return map[string]any{
		"protocol":             r.Protocol,
		"currency":             r.Currency,
		"smartContractAddress": r.SmartContractAddress,
		"companyUserId":        r.CompanyUserID,
		"receiptOrderId":       r.ReceiptOrderID,
		"currencyNumber":       r.CurrencyNumber.String(),
		"callBackUrl":          r.CallBackURL,
	}
}

// ReceiptConfirmRequest confirms a receipt order.
type ReceiptConfirmRequest struct {
	Protocol             string
	Currency             string
	SmartContractAddress string
	CompanyUserID        string
	ReceiptOrderID       string
}

func (r *ReceiptConfirmRequest) Method() string     { return http.MethodPost }
func (r *ReceiptConfirmRequest) Path() string       { return "/pay/v3/receipt/order/confirm" }
func (r *ReceiptConfirmRequest) Extra() http.Header { return nil }
func (r *ReceiptConfirmRequest) Query() url.Values  { return nil }
func (r *ReceiptConfirmRequest) Body() any {
	return map[string]any{
		"protocol":             r.Protocol,
		"currency":             r.Currency,
		"smartContractAddress": r.SmartContractAddress,
		"companyUserId":        r.CompanyUserID,
		"receiptOrderId":       r.ReceiptOrderID,
	}
}

File: rust/src/pay/receipt.rs
use rust_decimal::Decimal;
use crate::request::{ApiRequest, HttpMethod};
use super::model::{PayOrderAddressFixedResp, PayProtocolsResp, ReceiptOrderAddressResp, ReceiptOrderResp};

pub struct ReceiptProtocolsRequest {
    /// 1 = dynamic address, 6 = fixed address.
    pub receipt_type: i32,
}

impl Default for ReceiptProtocolsRequest {
    fn default() -> Self { Self { receipt_type: 1 } }
}

impl ApiRequest for ReceiptProtocolsRequest {
    type Response = Vec<PayProtocolsResp>;
    fn method(&self) -> HttpMethod { HttpMethod::Get }
    fn path(&self)   -> &str       { "/pay/v3/protocols" }
    fn query_params(&self) -> Vec<(String, String)> {
        vec![("type".into(), self.receipt_type.to_string())]
    }
}

pub struct ReceiptAddressRequest {
    pub protocol:               Option<String>,
    pub smart_contract_address: Option<String>,
    pub company_user_id:        Option<String>,
}

impl ApiRequest for ReceiptAddressRequest {
    type Response = PayOrderAddressFixedResp;
    fn method(&self) -> HttpMethod { HttpMethod::Get }
    fn path(&self)   -> &str       { "/pay/v3/receipt/address" }
    fn query_params(&self) -> Vec<(String, String)> {
        let mut p = vec![];
        if let Some(v) = &self.protocol               { p.push(("protocol".into(), v.clone())); }
        if let Some(v) = &self.smart_contract_address { p.push(("smartContractAddress".into(), v.clone())); }
        if let Some(v) = &self.company_user_id        { p.push(("companyUserId".into(), v.clone())); }
        p
    }
}

pub struct ReceiptOrdersRequest {
    pub receipt_order_ids: Option<String>,
    pub order_status:      Option<i32>,
    pub status:            Option<i32>,
}

impl ApiRequest for ReceiptOrdersRequest {
    type Response = Vec<ReceiptOrderResp>;
    fn method(&self) -> HttpMethod { HttpMethod::Get }
    fn path(&self)   -> &str       { "/pay/v3/receipt/orders" }
    fn query_params(&self) -> Vec<(String, String)> {
        let mut p = vec![];
        if let Some(v) = &self.receipt_order_ids { p.push(("receiptOrderIds".into(), v.clone())); }
        if let Some(v) = self.order_status        { p.push(("orderStatus".into(), v.to_string())); }
        if let Some(v) = self.status              { p.push(("status".into(), v.to_string())); }
        p
    }
}

pub struct ReceiptAddOrderRequest {
    pub protocol:               Option<String>,
    pub currency:               Option<String>,
    pub smart_contract_address: Option<String>,
    pub company_user_id:        Option<String>,
    pub receipt_order_id:       Option<String>,
    pub currency_number:        Option<Decimal>,
    pub call_back_url:          Option<String>,
}

impl ApiRequest for ReceiptAddOrderRequest {
    type Response = ReceiptOrderAddressResp;
    fn method(&self) -> HttpMethod { HttpMethod::Post }
    fn path(&self)   -> &str       { "/pay/v3/receipt/order/add" }
    fn body(&self) -> Option<serde_json::Value> {
        Some(serde_json::json!({
            "protocol":             self.protocol,
            "currency":             self.currency,
            "smartContractAddress": self.smart_contract_address,
            "companyUserId":        self.company_user_id,
            "receiptOrderId":       self.receipt_order_id,
            "currencyNumber":       self.currency_number.map(|d| d.to_string()),
            "callBackUrl":          self.call_back_url,
        }))
    }
}

pub struct ReceiptConfirmRequest {
    pub protocol:               Option<String>,
    pub currency:               Option<String>,
    pub smart_contract_address: Option<String>,
    pub company_user_id:        Option<String>,
    pub receipt_order_id:       Option<String>,
}

impl ApiRequest for ReceiptConfirmRequest {
    type Response = serde_json::Value;
    fn method(&self) -> HttpMethod { HttpMethod::Post }
    fn path(&self)   -> &str       { "/pay/v3/receipt/order/confirm" }
    fn body(&self) -> Option<serde_json::Value> {
        Some(serde_json::json!({
            "protocol":             self.protocol,
            "currency":             self.currency,
            "smartContractAddress": self.smart_contract_address,
            "companyUserId":        self.company_user_id,
            "receiptOrderId":       self.receipt_order_id,
        }))
    }
}
